app/splitwise.py

```diff
@@ -1,193 +1,205 @@
+        Adds expense to database
+        Fields 'name', 'sum', 'lender_id', 'event_token' are required
+        Fields 'id', 'datetime' will be ignored
+        @return: expense_id
+        """
+        expense.id = None  # will be filled after expense being added to db
+        expense.datetime = datetime.now()
+        return self.conn.save_expense_info(expense)
 
     def add_user_to_event(
         self,
         user_id: int,
-        event_id: str
+        event_token: str
     ) -> NoReturn:
-        self.conn.add_user_to_event(user_id, event_id)
+        self.conn.add_user_to_event(user_id, event_token)
 
-    def get_user_event(
+    def user_participates_in_event(
         self,
         user_id: int,
-        event_id: str
-    ) -> list:
-        return self.conn.get_user_event(user_id, event_id)
+        event_token: str,
+    ) -> bool:
+        return self.conn.user_participates_in_event(user_id, event_token)
 
     def get_expense(
         self,
-        expense_id: int
-    ) -> dict:
-        expense = self.conn.get_expense_info(expense_id)
-        return expense
+        expense_id: int,
+    ) -> Expense:
+        return self.conn.get_expense_info(expense_id)
 
     def add_debt(
-            self,
-            expense_id: int,
-            lender_id: int,
-            debtor_id: int,
-            sum_: float,
+        self,
+        debt: Debt,
     ) -> int:
-        return self.conn.save_debt_info(expense_id, lender_id, debtor_id, sum_)
+        return self.conn.save_debt_info(debt)
 
     def get_final_transactions(
         self,
-        token: str,
-    ) -> dict:
-        users = self.get_users_of_event(token)
-        users_map = dict()
-        users_balance = dict()
-        for user in users:
-            users_balance[user['id']] = 0.
-            users_map[user['id']] = user['name']
-
-        users_expenses = self.conn.get_event_expenses(token)
-        expenses_id = list()
-        for user_expenses in users_expenses:
-            expenses_id.append(user_expenses[0])
-            users_balance[user_expenses[1]] -= user_expenses[2]
-
-        debts = self.conn.get_debts_of_expenses(expenses_id)
-        for debt in debts:
-            users_balance[debt[0]] += debt[1]
-        lenders = [(-sum_, id_) for id_, sum_ in users_balance.items() if sum_ < 0]
-        debtors = [(sum_, id_) for id_, sum_ in users_balance.items() if sum_ > 0]
+        event_token: str,
+    ) -> Tuple[Dict, Dict]:
+        """
+        @param event_token: event token
+        @return: tuple (lenders_info, debtors_info)
+        """
+        users = self.get_users_of_event(event_token)
+        usernames = {user.id: user.name for user in users}
+        users_balance = defaultdict(lambda: {
+            'lent': 0,
+            'owed': 0,
+        })
+
+        event_expenses = self.conn.get_event_expenses(event_token)
+        for expense in event_expenses:
+            users_balance[expense.lender_id]['lent'] += expense.sum
```

Comment:
We have to be sure that expense amount is equal to sum of all debts related to the current expense